lib/_stream_pump.js

```diff
@@ -0,0 +1,191 @@
+    writable = false;
+    if (!readable) callback();
+  };
+
+  const onend = () => {
+    readable = false;
+    if (!writable) callback();
+  };
+
+  const onexit = (exitCode) => {
+    if (exitCode) {
+      callback(new Error('exited with error code: ' + exitCode));
+    } else {
+      callback();
+    }
+  };
+
+  const onclose = () => {
+    if (readable && !(rs && rs.ended))
+      return callback(new Error('premature close'));
+
+    if (writable && !(ws && ws.ended))
+      return callback(new Error('premature close'));
+  };
+
+  const onrequest = () =>
+    stream.req.on('finish', onfinish);
+
+  if (isRequest(stream)) {
+    stream.on('complete', onfinish);
+    stream.on('abort', onclose);
+    if (stream.req) onrequest();
+    else stream.on('request', onrequest);
+  } else if (writable && !ws) { // legacy streams
+    stream.on('end', onlegacyfinish);
+    stream.on('close', onlegacyfinish);
+  }
+
+  if (isChildProcess(stream)) stream.on('exit', onexit);
+
+  stream.on('end', onend);
+  stream.on('finish', onfinish);
+  if (opts.error !== false) stream.on('error', callback);
+  stream.on('close', onclose);
+
+  return () => {
+    stream.removeListener('complete', onfinish);
+    stream.removeListener('abort', onclose);
+    stream.removeListener('request', onrequest);
+    if (stream.req) stream.req.removeListener('finish', onfinish);
+    stream.removeListener('end', onlegacyfinish);
+    stream.removeListener('close', onlegacyfinish);
+    stream.removeListener('finish', onfinish);
+    stream.removeListener('exit', onexit);
+    stream.removeListener('end', onend);
+    stream.removeListener('error', callback);
+    stream.removeListener('close', onclose);
+  };
+}
+
+function destroyer(stream, reading, writing, _callback) {
+  let callbackCalled = false;
+  const callback = (err) => {
+    if (callbackCalled) return;
+
+    callbackCalled = true;
+
+    return _callback(err);
+  };
+  let closed = false;
+  stream.on('close', () => {
+    closed = true;
+  });
+
+  eos(stream, {readable: reading, writable: writing}, (err) => {
+    if (err) return callback(err);
+    closed = true;
+    callback();
+  });
+
+  var destroyed = false;
+  return (err) => {
+    if (closed) return;
+    if (destroyed) return;
+    destroyed = true;
+
+    if (isRequest(stream))
+      return stream.abort();
+      // request.destroy just do .end - .abort is what we want
+
+    if (isFn(stream.destroy)) return stream.destroy(err);
+
+    callback(err || new Error('stream was destroyed'));
+  };
+}
+
+const call = (fn) => fn();
+const callErr = (err) => (fn) => fn(err);
+const pipe = (from, to) => from.pipe(to);
+
+function pump(...streams) {
+  const callback = isFn(streams[streams.length - 1] || noop) &&
+    streams.pop() || noop;
+
+  if (Array.isArray(streams[0])) streams = streams[0];
+
+  if (streams.length < 2)
+    throw new Error('pump requires two streams per minimum');
+
+  let error;
+  const destroys = streams.map((stream, i) => {
```

Comment:
I would prefer not to close over `this`  here, as it might be the global `Stream`  object.